src/core/Akka.Cluster/Cluster.cs

```diff
@@ -333,28 +333,27 @@ public void JoinSeedNodes(IEnumerable<Address> seedNodes)
         /// </summary>
         /// <param name="seedNodes">TBD</param>
         /// <param name="token">TBD</param>
-        public Task JoinSeedNodesAsync(IEnumerable<Address> seedNodes, CancellationToken token = default(CancellationToken))
+        public Task JoinSeedNodesAsync(IEnumerable<Address> seedNodes, CancellationToken token = default)
         {
-            var completion = new TaskCompletionSource<NotUsed>(TaskCreationOptions.RunContinuationsAsynchronously);
-            
-            var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(token);
-            timeoutCts.CancelAfter(Settings.SeedNodeTimeout);
-            timeoutCts.Token.Register(() =>
+            if (_asyncJoinState != null)
             {
-                timeoutCts.Dispose();
-                completion.TrySetException(new ClusterJoinFailedException(
-                    $"Node has not managed to join the cluster using provided seed node addresses: {string.Join(", ", seedNodes)}."));
-            });
+                _log.Error("Another async cluster join is already in progress");
+                return _asyncJoinState.Task;
+            }
             
-            RegisterOnMemberUp(() =>
-            {
-                timeoutCts.Dispose();
-                completion.TrySetResult(NotUsed.Instance);
-            });
-
-            JoinSeedNodes(seedNodes);
-
-            return completion.Task.WithCancellation(token);
+            var nodes = seedNodes.ToList();
+            _asyncJoinState = new AsyncJoinState(
+                cluster: this, 
+                failException: new ClusterJoinFailedException(
+                    $"Node has not managed to join the cluster using provided seed node addresses: {string.Join(", ", nodes)}."), 
+                onComplete: () => _asyncJoinState = null,
```

Comment:
I wouldn't null the state out if we're going to return this value downstream.

Reply:
The state object is there as a marker that a join operation is already in progress, the most important part is the `Task` object from the `TaskCompletionSource` inside it.
Its totally fine that it is nulled.